hal/test-headers/ticker_api_tests.h

```diff
@@ -0,0 +1,138 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** \addtogroup hal_specification */
+/** @{*/
+
+#ifndef MBED_US_TICKER_API_TESTS_H
+#define MBED_US_TICKER_API_TESTS_H
+
+#include "device.h"
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \defgroup hal_specification_ticker Ticker
+ * Ticker specification and tests
+ *
+ * To run the microsecond ticker hal tests use the command:
+ *
+ *     mbed test -t <toolchain> -m <target> -n tests-mbed_hal-us_lp_ticker*
+ *
+ * # Defined behavior
+ * * The function ticker_init is safe to call repeatedly - Verified by test ::ticker_init_test
+ * * The function ticker_init resets the internal count and disables the ticker interrupt - Verified by test ::ticker_init_test
+ * * Ticker frequency is non-zero and counter is at least 8 bits - Verified by ::ticker_info_test
+ * * The ticker rolls over at (1 << bits) and continues counting starting from 0 - Verified by ::ticker_overflow_test
+ * * The ticker counts at the specified frequency +- 10% - Verified by ::ticker_frequency_test
+ * * The ticker increments by 1 each tick - Verified by ::ticker_increment_test
```

Comment:
It looks like this may be hard to prove. I understand that in current implementation us_ticker_read() returns us. This is to be changed by the vendors and in the new version they will define ticker clock frequency, ticker will be incremented always by 1 (e.g. currently on NRF51_DK we have 32KHz ticker clock and ticker value is incremented by 31) and us_ticker_read() will return pure ticker value. So in the new version on our NRF51_DK ticker will be incremented by one every 31 us. On this board read operation takes about 15 us, so we might be able to prove that ticker is incremented by one. But what with boards where the relation between CPU clock freq and ticker clock frequency is such that read operation is longer than delays between ticker increments. Currently on NUCLEO_F070RB for example ticker is incremented every 1us and read operation takes over one us.

Reply:
It may not be possible to comprehensively test this for the high speed tickers. The goal of this requirement is to get vendors to use the timer directly, rather than scaling the timer internally, as was done with the low power timer on most boards.

Worst case you could test this up to a point based on the required execution time of ticker_read. This may mean the execution time requirement on ticker_read be decreased from 20us to 10 or 15us to ensure ticks of a 32KHz clock can be read.

One possible way you could measure this indirectly is with the elapsed time of a delay loop. Assuming the loop executes in a consistent time then the the elapsed time should be accurate +-1us if there is no scaling involved. This could be run across a range of delay cycles to give more confidence that no scaling is done. This approach may turn out to be too complicated though. What do you think @mprse?
```c
uint32_t test_read_with_delay(uint32_t cycles)
{
    enter_critical();
    volatile uint32_t count = 1000;
    uint32_t start = us_ticker_read();
    while (count--);
    uint32_t stop = us_ticker_read();
    exit_critical();
}

```

Reply:
Lets assume that ticker is incremented every 1 us by 1 and read operation takes:
1) less than 1 us - in this case we can easily prove that ticker is incremented by 1 (not scaled).
2) more than 1 us, lets assume 10 us. Now when we would perform reads in sequence, then the difference between read tick counts should be 10 +/- 1 ticks. Now if we use provided while loop with cycles set to 10 and measure N ticks, then for 100 cycles we should have 10*N +/-1 tick, for 1000 cycles 100*N +/-1 tick.
In case when ticker would be increment by 5 ticks (scaled), then in the same scenario we would get 10 cycles : N ticks, 100 cycles : 10*N +/-5, 1000 cycles : 100*N +/-5. We might now detect that the counted value is not in rage +/-1 tick, but still we might not detect that ticker is incremented by 5 if in all cases we read exact values (10*N, 100*N).

On the other hand I think that frequency test using PC might be sufficient for this case. If we measure using PC number of ticker ticks during one second and validate this against ticker clock frequency then we prove that ticker is not scaled. E.g. Currently on NRF51_DK we have 32KHz ticker clock and ticker is incremented by 31 (scaled). So counted number of ticks will be about 1000000 and we expect about 32 000.

Reply:
I don't understand how measuring using a PC would help detect single ticks. If a 32KHz clock is scaled up internally to 1MHz then it will appear that the frequency is correct - over the course of 1s the tick value reported will be 1000000 +- 31.

As for point 2 mentioned above, I agree. With a fixed test delays a scaled frequency may not be detected. If you sweep the delay (delay = 100 +1 * sweep_cycle), then you should be able to detect any frequency scaling, as long as one cycle of the loop takes less than 1 tick.